samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java

```diff
@@ -0,0 +1,128 @@
+import java.util.concurrent.CompletableFuture;
+import org.apache.samza.SamzaException;
+import org.apache.samza.context.Context;
+import org.apache.samza.table.remote.TableReadFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Single;
+import rx.SingleSubscriber;
+
+
+public class CouchbaseTableReadFunction<V> extends BaseCouchbaseTableFunction<V>
+    implements TableReadFunction<String, V> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CouchbaseTableReadFunction.class);
+
+  protected final Class<? extends Document<?>> documentType;
+
+  public CouchbaseTableReadFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
+    super(bucketName, clusterNodes, valueClass);
+    documentType = JsonObject.class.isAssignableFrom(valueClass) ? JsonDocument.class : BinaryDocument.class;
+  }
+
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info(String.format("Read function for bucket %s initialized successfully", bucketName));
+  }
+
+  @Override
+  public CompletableFuture<V> getAsync(String key) {
+    Preconditions.checkNotNull(key);
+    CompletableFuture<V> future = new CompletableFuture<>();
+    Single<? extends Document<?>> singleObservable =
+        bucket.async().get(key, documentType, timeout, timeUnit).toSingle();
+    singleObservable.subscribe(new SingleSubscriber<Document>() {
+      @Override
+      public void onSuccess(Document document) {
+        if (document != null) {
+          if (document instanceof JsonDocument) {
+            future.complete((V) document.content());
+          } else {
+            BinaryDocument binaryDocument = (BinaryDocument) document;
+            ByteBuf buffer = binaryDocument.content();
+            try {
+              byte[] bytes;
+              if (buffer.hasArray() && buffer.arrayOffset() == 0 && buffer.readableBytes() == buffer.array().length) {
+                bytes = buffer.array();
+              } else {
+                bytes = new byte[buffer.readableBytes()];
+                buffer.readBytes(bytes);
+              }
+              future.complete(valueSerde.fromBytes(bytes));
+            } catch (Exception e) {
+              future.completeExceptionally(
+                  new SamzaException(String.format("Failed to deserialize value of key %s with given serde", key), e));
+            } finally {
+              ReferenceCountUtil.release(buffer);
+            }
+          }
+        } else {
+          future.complete(null);
+        }
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        if (throwable instanceof NoSuchElementException) {
+          future.complete(null);
+        } else {
+          future.completeExceptionally(new SamzaException(String.format("Failed to get key %s", key), throwable));
+        }
+      }
+    });
+    return future;
+  }
+
+  @Override
+  public boolean isRetriable(Throwable throwable) {
```

Comment:
Don't need this method.

Reply:
I decided to keep it because isRetriable() is not required by any interface in `BaseCouchbaseTableFunction`. To avoid someone changing the name of isRetriable() in the base class, I added this call here in the child class. Does it make sense?

Reply:
If someone changes it, it won't compile.

Reply:
ok